### PR TITLE
Implement P0952R2: 'A new specification for `std::generate_canonical`'

### LINKED ISSUE #1074: <random>: uniform_real_distribution(a, b) produces outputs outside [a, b)
**Describe the bug**
> Produces random floating-point values i, uniformly distributed on the interval [a, b)

*[cppreference.com](https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution)*

The below code sample demonstrates how a default-constructed `uniform_real_distribution<float>` (a=0, b=1) generates a `1.0f` if provided a random number with all bits set:

**Command-line test case**
```
#include <cassert>
#include <cstdint>
#include <limits>
#include <random>

struct max_rng
{
	using result_type = std::uint32_t;
	result_type min() const { return 0; }
	result_type max() const { return std::numeric_limits<result_type>::max(); }
	result_type operator()() { return max(); }
};

int main()
{
	max_rng rng;
	std::uniform_real_distribution<float> iid;

	assert(iid(rng) < 1.0f);

	return 0;
}
```

**Expected behavior**
Regardless of the input received from the RNG, `uniform_real_distribution` should never produce `b` as the interval is supposed to be right-*exclusive*. GCC passes this test.

**STL version**
Microsoft Visual Studio Community 2019
Version 16.6.4

**Additional context**
The error seems to originate from [`generate_canonical`](https://github.com/microsoft/STL/blob/c10ae01b4d9508eed9d5f059a120ee7223b6ac12/stl/inc/random#L261), which, when passed in the above code, simplifies down to:

```
return (float) UINT_MAX / (UINT_MAX + 1.0f);
```
which, mathematically speaking, *should* be less than 1, but in practice gets rounded to `1.0f` due to limited floating point precision. `generate_canonical` is flawed in other regards: using a normalizing/scaling division leads to non-uniformly distributed values. Both of these issues could be avoided by generating random floats the "canonical" way, which is to set the bits of the mantissa directly (and simply throw away left-over entropy) along the lines of:
```
// @rnd: uniformly distributed 32bit unsigned int.
return (rnd >> 8) / 164233216.0f
```
which not only produces values strictly from [0, 1) for any value of `rnd`, but also makes sure that all values are equally far spaced apart leading to a perfect uniform distribution.

Also tracked by DevCom-110322 and Microsoft-internal VSO-253526 / [AB#253526](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/253526) .

### LINKED ISSUE #4169: P0952R2 A New Specification For `generate_canonical()`
WG21-P0952R2 A New Specification For `generate_canonical()`

No feature-test macro is mentioned.

This technically isn't a Defect Report, but it's a behavioral change that resolves an LWG issue and avoids the correctness-damaging behavior of returning `1.0`, so we want to implement this unconditionally and **WILL** accept a PR for this in advance of opening up the floodgates to arbitrary C++26 features at some point in the future.

### PR DESCRIPTION
**Pull request by MattStephanson**
Fixes #1074. Fixes #4169.

### CODE REVIEW

File: stl/inc/random
Comment (MattStephanson) [inline stl/inc/random:266]:
```suggestion
    uint64_t _Xx     = 0; // $x$ in [rand.util.canonical]/1.5

```

File: stl/inc/random
Comment (MattStephanson) [inline stl/inc/random:318]: Need to copy new implementation to `_Nrand_impl` and audit tr1 tests.

**Comment by MattStephanson** [comment]
The failed test checks the internal helper function `_Generate_canonical_iterations`, which I've removed and replaced with a version that calculates more parameters at compile time. I could either adapt the test to the new implementation or just remove the test entirely. I'm leaning toward removing it because (1) testing internal function is generally undesirable and (2) the test was testing a naive implementation of Standard wording that no long exists ($k = \mathrm{max}(1, \lceil b / log_2 R\rceil)$) after WG21-P0952. I'd welcome any comment on this point.

File: stl/inc/random
Comment (StephanTLavavej) [reply stl/inc/random:266]: When citing a paragraph, it's important to include the WP number, because paragraphs are frequently renumbered. (Stable names by themselves are stable enough that they don't need WP numbers.)

**Comment by StephanTLavavej** [comment]
> I'm leaning toward removing it because (1) testing internal function is generally undesirable and (2) the test was testing a naive implementation of Standard wording that no long exists

Removing it sounds good to me.

File: stl/inc/random
Comment (StephanTLavavej) [inline stl/inc/random:1]: Also remember to list the "cleaned up" paper title in `yvals_core.h` to reflect how it's implemented.

File: stl/inc/random
Comment (frederick-vs-ja) [inline stl/inc/random:272]: This function is not supposed to be called at run time. Is it valuable to make it `consteval` since C++20?
But conditional compilation is possibly undesired as #3935 removed the internal macro `_CONSTEVAL`.

File: stl/inc/random
Comment (StephanTLavavej) [inline stl/inc/random:423]: Let's also comment this as TRANSITION, being for non-integral tr1.

File: stl/inc/random
Comment (StephanTLavavej) [inline stl/inc/random:422]: Style: We conventionally prefer positive tests unless there's a clarity advantage otherwise (it doesn't even need to be a huge advantage, any small reason will do). Here, I see no reason - the order Standard, integral tr1, non-integral tr1 would be fine, and sorted in order of ascending wackiness, so let's avoid the negation.

File: tests/std/tests/P0952R2_new_generate_canonical/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P0952R2_new_generate_canonical/test.cpp:122]: This test is already `using namespace std;`, but actually we can just use range-for here. Occurs below.

File: tests/std/tests/P0952R2_new_generate_canonical/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P0952R2_new_generate_canonical/test.cpp:126]: Should include `<cmath>`.

File: tests/std/tests/P0952R2_new_generate_canonical/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P0952R2_new_generate_canonical/test.cpp:11]: I believe `<utility>` is unused here - please meow if I'm wrong.

File: stl/inc/random
Comment (StephanTLavavej) [reply stl/inc/random]: Good question. As this is purely internal and used in only one place, I don't think conditionally using `consteval` is worth the effort.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks, this is awesome! :heart_eyes_cat:

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

File: stl/inc/random
Comment (statementreply) [inline stl/inc/random:334]: `_Sx` doesn't need to be wider than `_Result_uint_type` when `_Rx` is a power of 2, as is noted in https://eel.is/c++draft/rand.util.canonical#7 . The $i$-th random value could be shifted by $i \log_2 R - \log_2 x$ bits instead (the 0th random value is shifted right).

File: stl/inc/random
Comment (StephanTLavavej) [reply stl/inc/random:285]: That's a good question. Since this would involve `generate_canonical` being given `long double` as a template argument, it wouldn't affect all 80-bit `long double` users, only those who are specifically using `generate_canonical` with it. I think we should go ahead with the code as-is, and if someone complains we're regressing them, then think about altering the code. (Our quasi-support continues to be a thorn in our side...)

File: stl/inc/random
Comment (StephanTLavavej) [reply stl/inc/random:334]: Thanks, I've filed #4750 for followup.

**Comment by MattStephanson** [comment]
Thanks Stephan! I have a few follow-up comments, but I'm at work right now
and can't sign in to GitHub. Would you mind keeping this open until I can
respond?

**Comment by StephanTLavavej** [comment]
Sure thing, I'll hold off on merging this until you can respond. If I messed up something that should be fixed before merging instead of addressed in a followup, I can easily pull this from the merge batch for rework. Thanks! :heart_eyes_cat:

**Comment by StephanTLavavej** [comment]
Looks like there are `/clr:pure` timeouts (after 10 minutes) for `GH_001017_discrete_distribution_out_of_range`, possibly involving the 128-bit dependency, so I'll need to investigate that and push a commit later.

File: stl/inc/random
Comment (MattStephanson) [reply stl/inc/random]: I was definitely getting a warning while developing the test code in a VS sandbox project that I have. I'll see if I can repro.

File: stl/inc/random
Comment (MattStephanson) [reply stl/inc/random]: FWIW, this was intentional to reduce the number different instantiations. As you said, optimization isn't a big concern for the tr1 path, so I figured the throughput benefit of fewer instantiations would be preferable. I could be completely mistaken about the tradeoffs here, though.

File: stl/inc/random
Comment (MattStephanson) [reply stl/inc/random:285]: Yes, it will fail for x87 extended. It can be easily fixed by special-casing the 64-bit case, but is there a clear, `constexpr`-friendly way to write 2^-64 in pre-17 mode?

File: stl/inc/random
Comment (StephanTLavavej) [reply stl/inc/random]: Yeah, I think you're right. I'll switch it back, thanks.

File: stl/inc/random
Comment (StephanTLavavej) [reply stl/inc/random]: Thanks - if you find a repro, we can look into the root cause or restore the suppression.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for implementing this major overhaul of central `<random>` machinery! :gear: :hammer_and_wrench: :green_heart:

**Pull request merged**
